openvpn/middlewares/server/auth/middleware.go

```diff
@@ -19,41 +19,89 @@ package auth
 
 import (
 	"strings"
+	"sync"
 
 	"github.com/mysteriumnetwork/go-openvpn/openvpn/log"
 	"github.com/mysteriumnetwork/go-openvpn/openvpn/management"
 	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server"
 )
 
-type middleware struct {
-	// TODO: consider implementing event channel to communicate required callbacks
-	credentialsValidator CredentialsValidator
-	commandWriter        management.CommandWriter
-	currentEvent         server.ClientEvent
-}
+// ClientEventCallback is called when state of each Openvp client changes.
+type ClientEventCallback func(event server.ClientEvent)
+
+// Middleware is able to process client auth events, exposes client control API.
+//
+// The OpenVPN server should have been started with the
+// --management-client-auth directive so that it will ask the management
+// interface to approve client connections.
+type Middleware struct {
+	commandWriter management.CommandWriter
+	currentEvent  server.ClientEvent
 
-// CredentialsValidator callback checks given auth primitives (i.e. customer identity signature / node's sessionId)
-type CredentialsValidator func(clientID int, username, password string) (bool, error)
+	listenersMu sync.RWMutex
+	listeners   []ClientEventCallback
```

Comment:
When are listeners removed? This could lead to memory leaks if they are never removed.

Reply:
It's same for all middlewares. We simple don't reuse created instances. They are created/destroyed with server start/stop